### ISSUE TITLE
‘test_proto2::test_proto2’ has not been declared

### ISSUE THREAD

**Issue by Fisher1006** [tomas-abrahamsson/gpb#168]
Hey, I have got error when I try to compile the nif.cc file

I make two proto files:

test_proto2.proto:

> syntax = "proto3";
>
>import "proto1.proto";
>
>package test_proto2;
>
>message m_test {
>    int64 id = 1;
>    proto1.p_kv kv = 2;
>}

proto1.proto:

>syntax = "proto3";
>
>package proto1;
>
>message p_kv {
>    uint64 id = 1;
>    uint64 val = 2;
>}

And I execute the cmd:

>protoc-erl -o-erl ErlDir  -o-hrl IncDir -nif -pkgs -o-nif-cc CCDir -il test_proto2.proto
>
>protoc --cpp_out=CCDir --proto_path=. ./*.proto
>
>ccache g++ -g -std=c++11 -fPIC -shared -O3 -Wall -I/usr/local/otp/20.3/erts-9.3/include \
CCDir/test_proto2.pb.cc CCDir/test_proto2.nif.cc -o CCDir/test_proto2.nif.so  -Wl,--whole-archive /usr/local/lib/libprotobuf.a -Wl,--no-whole-archive`

If I remove the  `-pkgs` option,  it will exit on cmd `protoc-erl ...` and error:`in msg m_test, field kv: undefined reference  proto1.p_kv`

If I add the `-pkgs` option, it will exit on cmd `g++` and error: `error: ‘test_proto2::test_proto2’ has not been declared`

I searched the stackoverflow and got a similar question: [about namespace](https://stackoverflow.com/questions/6994506/google-protocol-buffers-c-error-protobuf-does-not-name-a-type), and looked the related  issues but still didn't solve the problem, please give me some advice

Thanks

**Comment by tomas-abrahamsson** [comment]
Hmm.. looks like a bug in gpb when it generates the `.nif.cc` file. It generates `::test_proto2::test_proto2` but it should have been `::test_proto2`. The same goes for `::test_proto2::proto1` which should be just `::proto1`. I'll look into it.

**Comment by tomas-abrahamsson** [comment]
I have a preview of a fix on a temporary branch `nif-pkgs`, if you would like to try it out. I will force-push to it later on when I have added unit tests.

**Comment by Fisher1006** [comment]
Thank you very much, I will try it out today and feedback

**Comment by Fisher1006** [comment]
I copied the gpb_gen_nif.erl and gpb_lib.erl from the nif-pkgs branch, it works well for now

**Comment by Fisher1006** [comment]
another question：
Am I must use `pkgs` option if I want to use import features? 
And if I use `pkgs` option, the proto record definitions will prefix PackageName, it will look like `PackageName.Record#{}` , how can I use like `Record#{}` pattern?

**Comment by tomas-abrahamsson** [comment]
> I must use pkgs option if I want to use import features?

If they are in different `package`s, then yes.  This is since names can otherwise collide, for example if there is `pkg.a.Msg` and `pkg.b.Msg`, then it is not possible to refer to `Msg` without package (and the package name is needed to resolve references such as for the field `proto1.p_kv kv = 2;` in your example).  But if the `import`ed proto has the same `package`, then names will not collide, then the `-pkgs` option is not needed.

> And if I use pkgs option, the proto record definitions will prefix PackageName, it will look like PackageName.Record#{} , how can I use like Record#{} pattern?

There is a renaming option, `{rename,{msg_fqname,base_name}}` option that strips away the package name after parsing. There will be an error if this results in a name collision. Follow this link for documentation on the [rename option](https://hexdocs.pm/gpb/gpb_compile.html#option-rename). 

However, the `rename` option is not available from the command line, and the nif generation is currently not aware of it either. Not even on the `nif-pkgs` branch.

I should fix the nif generation, I just need some more time for that. I think the same holds for making the renaming options available via the command line, eventually. Generally, these things are driven by demand, from me or from others.

**Comment by Fisher1006** [comment]
Thanks for reply so quick，because of the prefix name, I write all the protos to one proto file, and when I compile the  cc file to .so it cost too much time, it has been more than ten minutes and still not finish. Is this normal or did I do something wrong, I use the cmd mentioned before:

> ccache g++ -g -std=c++11 -fPIC -shared -O3 -Wall -I/usr/local/otp/20.3/erts-9.3/include 
CCDir/test_proto2.pb.cc CCDir/test_proto2.nif.cc -o CCDir/test_proto2.nif.so -Wl,--whole-archive /usr/local/lib/libprotobuf.a -Wl,--no-whole-archive`

by the way, I look at the document of Erlang nif and maybe don't digger too depth and I don't know why I need A.pb.cc after I have the  A.nif.cc, I tried compile the .so file just use A.nif.cc and it occur error `:undefined symbol ` when I load the module. Can you give me some notice so I can search it self ?

**Comment by tomas-abrahamsson** [comment]
The A.nif.cc contains the conversion to Erlang data types, While the A.pb.cc is generated by the Google protoc and packs and unpacks wire bytes to C++ objects.

I'll merge the nif-pkgs branch as is and make a new release, then. And fix renaming for nif and command line options in a later stage.

**Comment by tomas-abrahamsson** [comment]
I just pushed 4.9.1, which includes the `nif-pkgs` branch. (I will delete that temporary branch soon.) 

I still intend to fix so that nif generation takes renaming into consideration, and make renaming options available form the command line. Do you wish to keep this ticket open until that is done, or maybe you do not need those fixes/features so much any more (if so, we can close this ticket and I will have somewhat lower prio for those fixes)?

**Comment by Fisher1006** [comment]
I will keep it open, I have worked for a month to use the gpb. I hope the problems will be solved. To use the gpb, I have to solve the import problem and hot update problem which still not solve perfectly. 

As a result of prefix packagename, I tried to write all proto to single file, and the nif.cc file size is 5.47M, pb.cc file size is 18M, I compile the.so lib for half an hour but it just keep waiting. I don’t know why, is it right?

Thanks again

**Comment by tomas-abrahamsson** [comment]
Ok, no problems. I'll look into it. (Meanwhile, there will be some extra releases of things that are already pending.)

About the long compilation times, I don't really know, I suppose since the files are really huge, it may be possible it takes such long time. Especially if maybe the computer is out of memory and swapping, I suppose. I've noticed earlier that `map<_,_>` fields (which uses C++ templates) tend to increase compilation times.

____
Also, don't know if it applies to your platform or not, but for me, I could not link with the `libprotobuf.a`. The linker complained a lot and said:
```
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libprotobuf.a(bytestream.o): relocation R_X86_64_PC32 against symbol `_ZTVN6google8protobuf7strings20GrowingArrayByteSinkE' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libprotobuf.a(common.o): relocation R_X86_64_PC32 against symbol `stderr@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
...58 more similar lines...
```
I tried removing `-fPIC -shared` but that didn't work either, so I instead I just used `-lprotobuf` instead of the `-Wl,--whole-archive /.../ibprotobuf.a -Wl,--no-whole-archive` options. (I'm on a linux x86_64 debian unstable, with gcc/g++ 8.3.0-7 and ld version 2.32.51.20190707)

**Comment by Fisher1006** [comment]
Well the problem with link to protobuf I met also before. It seems that need to install protobuf with configure disable-share which means install protobuf static and use the libproto.a, I don’t know clearly why. But I got it worked after install protobuf static and use the libproto.a

**Comment by Fisher1006** [comment]
An advice, when we use the gpb, we just need the encode, decode and verify functions, the merge and introspect functions generated a lot code which we don't need. So I modified the gpb_compile.erl and add two options to control if need generated merge functions and introspect functions, I think maybe it can be provide by gpb to let others to custom decide to generate these functions whether or not.

**Comment by tomas-abrahamsson** [comment]
Ah, I see (about the way to compile `libprotobuf.a`)

Yes, this may be a useful addition, I'll consider it (about options to skip code for mergers and introspection)

**Comment by tomas-abrahamsson** [comment]
I force pushed to the `nif-pkgs` branch. It adds renamings support to the nif generation, and it adds a `-rename` command line option. Here is how to use the command line option (excerpt from the `protoc-erl --help`)
```
   -rename  What:How
       What:
         pkg_name       Modify the package name. Useful
                        with the -pkgs option.
         msg_name       Modify message names, but not the package
                        partof them
         msg_fqname     Modify message names, including their
                        package part. Useful with the -pkgs option.
         group_name     Group names.
         group_fqname   Group names including their package parts.
         service_name   Service names.
         service_fqname Service names including their package parts.
         rpc_name       The RPC name.
         msg_typename   Erlang type names for messages and groups.
         enum_typename  Erlang type names for enums.
       How:
          prefix=Prefix        Prepend the Prefix.
          suffix=Suffix        Append the Suffix.
          lower_case           Example: from MsgName to msgname
          snake_case           Example: from MsgName to msg_name
          dots_to_underscores  Example: from Msg.Sub to Msg_Sub
          base_name            Example: from Domain.Pkg.Msg to Msg
          proto=Proto:prefix=Prefix[,...]
                               For each message belonging the the
                               proto Proto, without the `.proto'
                               suffix, prepend Prefix. This only
                               works for msg_name and msg_fqname.
       It is possible to specify more than one -rename option,
       and they are applied in the order specified.
```

In this case, I might be safest not to pick files individually, or if you do, please take care to include all relevant files. The recently released 4.9.2 adds a new source file, `src/gpb_defs.erl`. The `nif-pkgs` is now based on 4.9.3.

**Comment by Fisher1006** [comment]
Thanks very much ! ! ! It works，just I didn't read the document carefully and mix the msg_name with msg_fqname waste some time to read the source code.  And thanks for the notice I replaced the whole 
 gpb project. The problem is solved, I think the issue can be close now.

**Comment by tomas-abrahamsson** [comment]
Super, you are welcome! I'll merge and release. I forgot to mention it, but I updated the branch earlier with two options `-no-gen-mergers` and `-no-gen-introspect`. They are included in the 3ced4c5 and 0b60947 commits.

**Comment by tomas-abrahamsson** [comment]
Just released 4.10.0 including the above, so I'll close this ticket.
